clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -506,18 +504,34 @@ static CallInst *CreateBuiltinCallWithAttr(CodeGenFunction &CGF,
 
 // Emit a simple mangled intrinsic that has 1 argument and a return type
 // matching the argument type. Depending on mode, this may be a constrained
-// floating-point intrinsic.
+// or an fpbuiltin floating-point intrinsic.
 static Value *emitUnaryMaybeConstrainedFPBuiltin(CodeGenFunction &CGF,
                                 const CallExpr *E, unsigned IntrinsicID,
                                 unsigned ConstrainedIntrinsicID) {
   llvm::Value *Src0 = CGF.EmitScalarExpr(E->getArg(0));
-  if (CGF.CGM.getCodeGenOpts().FPAccuracy) {    
-    Function *Func =
-        CGF.CGM.getIntrinsic(ConstrainedIntrinsicID, Src0->getType());
-    return CreateBuiltinCallWithAttr(CGF, Func, {Src0},
+  if (CGF.CGM.getCodeGenOpts().FPAccuracy) {
+    unsigned BuiltinID = CGF.getCurrentBuiltinID();
+    StringRef Name = CGF.CGM.getContext().BuiltinInfo.getName(BuiltinID);
+    Function *Func;
+    // Use fpbuiltin intrinsic only when needed.
+    bool IsBuiltin = false;
```

Comment:
The use of "builtin" here could be confusing, since this function is being called in the context of handling clang builtins. I was going to suggest that using "fpbuiltin" would be better, but it's only marginally better since a clang builtin for a math function is a kind of floating-point builtin. I can't think of a better way to make the distinction concisely though. Maybe HasAccuracyRequirement?